include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -1078,29 +1080,107 @@ void RunInParallelSection(ThreadPoolParallelSection &ps,
 // Run a single parallel loop _without_ a parallel section.  This is a
 // special case of RunInParallelSection, avoiding code paths for
 // handing off multiple loops to the pool of workers.
-
 void RunInParallel(std::function<void(unsigned idx)> fn, unsigned n, std::ptrdiff_t block_size) override {
   profiler_.LogStartAndCoreAndBlock(block_size);
-  PerThread *pt = GetPerThread();
+  PerThread* pt = GetPerThread();
   ThreadPoolParallelSection ps;
   StartParallelSectionInternal(*pt, ps);
 
-  // Summon workers to run the function (n is the desired maximum
-  // degree of parallelism, including the main thread).  Unlike the
-  // multi-loop RunInParallelSection, this single-loop worker can run
-  // fn directly without needing to receive it via ps.current_loop.
-  SummonWorkers(*pt, ps, n, fn);
-  profiler_.LogEndAndStart(ThreadPoolProfiler::DISTRIBUTION);
+  // index of thread that dispatch work to all other threads
+  int dispatch_q_idx = -1;
+  int dispatch_w_idx = -1;  // index of enqueued work
+  std::atomic<bool> dispatch_done{false};
+  std::atomic<bool> dispatch_work_done{false};
 
-  // Run work in the main thread
-  fn(0);
+  Task worker_fn = [&]() {
+    fn(ps.worker_idx.fetch_add(1, std::memory_order_relaxed));
+    ps.tasks_finished.fetch_add(1, std::memory_order_release);
+  };
+
+  std::function<int(int, Task&, bool)> enqueue_fn = [&, this](int q_idx, Task& w, bool logging) {
+    WorkerData& td = worker_data_[q_idx];
+    Queue& q = td.queue;
+    unsigned w_idx;
+    Task t = q.PushBackWithTag(w, pt->tag, w_idx);
+    if (t) {
+      return -1;
+    } else {
+      td.EnsureAwake();
+      if (logging) {
+        ps.tasks.push_back({q_idx, w_idx});
+      }
+      return static_cast<int>(w_idx);
+    }
+  };
+
+  unsigned current_dop = static_cast<unsigned>(ps.tasks.size()) + 1;
+  unsigned extra_needed = n > current_dop ? n - current_dop : 0;
+  std::vector<unsigned> good_hints, alt_hints;
+
+  Task dispatch_task = [&]() {  // dispatch work to other threads
```

Comment:
Could you add some comments here on the overall approach, and timeline of operations in (1) the main thread, (2) the first thread running dispatch work, and (3) the additional threads helping in the loop.  Maybe ASCII-art :)

This is important for correctness -- if I am following the logic correctly, we need the main thread to remain in this function (hence for the dispatch thread to be OK accessing the local variables it captures).

Reply:
Comments added ahead of function stating behaviors of main, dispatching and all other threads.